File: adapter/scripts/codelldb/webview.py
from typing import Any, Optional
from . import interface
from .event import Event
import lldb

view_id = 0


class Webview:
    '''A simplified interface for [webview panels](https://code.visualstudio.com/api/references/vscode-api#WebviewPanel).'''

    def __init__(self, debugger_id):
        global view_id
        view_id += 1
        self.id = view_id
        self.debugger_id = debugger_id
        self._on_did_receive_message = Event()
        self._on_did_dispose = Event()
        interface.on_did_receive_message.add(self._message_handler)

    def _message_handler(self, message):
        if message.get('id', None) == self.id:
            message_type = message.get('message', None)
            if message_type == 'webviewDidReceiveMessage':
                self._on_did_receive_message.emit(message.get('inner', None))
            elif message_type == 'webviewDidDispose':
                self._on_did_dispose.emit(message.get('inner', None))

    def __del__(self):
        interface.on_did_receive_message.remove(self._message_handler)

    def dispose(self):
        '''Destroy webview panel.'''
        interface.send_message(self.debugger_id, dict(message='webviewDispose', id=self.id))

    def set_html(self, html: str):
        '''Set HTML contents of the webview.'''
        interface.send_message(self.debugger_id, dict(message='webviewSetHtml', id=self.id, html=html))

    def reveal(self,  view_column: Optional[int] = None, preserve_focus: bool = False):
        '''Show the webview panel in a given column.'''
        interface.send_message(self.debugger_id, dict(message='webviewReveal', id=self.id,
                                                      viewColumn=view_column, preserveFocus=preserve_focus))

    def post_message(self, message: Any):
        '''Post a message to the webview content.'''
        interface.send_message(self.debugger_id, dict(message='webviewPostMessage', id=self.id, inner=message))

    @property
    def on_did_receive_message(self) -> Event:
        '''Fired when webview content posts a new message.'''
        return self._on_did_receive_message

    @property
    def on_did_dispose(self) -> Event:
        '''Fired when the webview panel is disposed (either by the user or by calling dispose())'''
        return self._on_did_dispose

File: extension/webview.ts
import {
    window, debug, DebugSession, DebugSessionCustomEvent, WebviewPanel, ViewColumn
} from "vscode";
import { Dict, DisposableSubscriber } from './novsc/commonTypes';

interface DebuggerPanel extends WebviewPanel {
    preserveOrphaned: boolean
}

export class WebviewManager extends DisposableSubscriber {
    sessionPanels: Dict<Dict<DebuggerPanel>> = {};

    constructor() {
        super();
        this.subscriptions.push(debug.onDidTerminateDebugSession(this.onTerminatedDebugSession, this));
        this.subscriptions.push(debug.onDidReceiveDebugSessionCustomEvent(this.onDebugSessionCustomEvent, this));
    }

    onTerminatedDebugSession(session: DebugSession) {
        if (session.type == 'lldb') {
            let panels = this.sessionPanels[session.id];
            if (panels) {
                for (let panel of Object.values(panels)) {
                    if (!panel.preserveOrphaned)
                        panel.dispose();
                }
            }
        }
    }

    onDebugSessionCustomEvent(e: DebugSessionCustomEvent) {
        if (e.session.type == 'lldb') {
            if (e.event == '_pythonMessage') {
                if (e.body.message == 'webviewCreate') {
                    this.createWebview(e.session, e.body);
                } else if (e.body.message == 'webviewDispose') {
                    this.sessionPanels[e.session.id][e.body.id].dispose();
                } else if (e.body.message == 'webviewSetHtml') {
                    this.sessionPanels[e.session.id][e.body.id].webview.html = e.body.html;
                } else if (e.body.message == 'webviewReveal') {
                    this.sessionPanels[e.session.id][e.body.id].reveal(e.body.viewColumn, e.body.preserveFocus)
                } else if (e.body.message == 'webviewPostMessage') {
                    this.sessionPanels[e.session.id][e.body.id].webview.postMessage(e.body.inner);
                }
            }
        }
    }

    createWebview(session: DebugSession, body: any) {
        let view_id = body.id;
        let panel = <DebuggerPanel>window.createWebviewPanel(
            'codelldb.webview',
            body.title || session.name,
            {
                viewColumn: body.viewColumn != null ? body.viewColumn : ViewColumn.Active,
                preserveFocus: body.preserveFocus
            },
            {
                enableFindWidget: body.enableFindWidget,
                enableScripts: body.enableScripts,
                retainContextWhenHidden: body.retainContextWhenHidden
            }
        );
        panel.webview.onDidReceiveMessage(e => {
            session.customRequest('_pythonMessage', { message: 'webviewDidReceiveMessage', id: view_id, inner: e });
        });
        panel.onDidDispose(() => {
            delete this.sessionPanels[session.id][view_id];
            session.customRequest('_pythonMessage', { message: 'webviewDidDispose', id: view_id });
        });
        if (body.html)
            panel.webview.html = body.html;
        panel.preserveOrphaned = body.preserveOrphaned

        let panels = this.sessionPanels[session.id];
        if (panels == undefined) {
            panels = {};
            this.sessionPanels[session.id] = panels;
        }
        panels[view_id] = panel;
    }
}

